File: mailblastr-dotnet/src/Mailblastr/Models/Domains.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Mailblastr;

/// <summary>One DNS record a domain needs (DKIM / SPF / DMARC / Tracking / Receiving MX).</summary>
public class DomainRecord
{
    /// <summary><c>DKIM</c> | <c>SPF</c> | <c>DMARC</c> | <c>Tracking</c> | <c>Receiving MX</c>.</summary>
    [JsonPropertyName("record")]
    public string Record { get; set; } = null!;

    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;

    /// <summary><c>CNAME</c> | <c>MX</c> | <c>TXT</c> | <c>CAA</c>.</summary>
    [JsonPropertyName("type")]
    public string Type { get; set; } = null!;

    [JsonPropertyName("ttl")]
    public string? Ttl { get; set; }

    [JsonPropertyName("status")]
    public string? Status { get; set; }

    [JsonPropertyName("value")]
    public string Value { get; set; } = null!;

    [JsonPropertyName("priority")]
    public int? Priority { get; set; }
}

/// <summary>Sending/receiving capabilities of a domain.</summary>
public class DomainCapabilities
{
    /// <summary><c>enabled</c> | <c>disabled</c>.</summary>
    [JsonPropertyName("sending")]
    public string? Sending { get; set; }

    /// <summary><c>enabled</c> | <c>disabled</c>.</summary>
    [JsonPropertyName("receiving")]
    public string? Receiving { get; set; }
}

/// <summary>Capabilities to enable on create/update, e.g. <c>Receiving = "enabled"</c>.</summary>
public class DomainCapabilitiesOptions
{
    /// <summary><c>enabled</c> | <c>disabled</c>.</summary>
    [JsonPropertyName("receiving")]
    public string? Receiving { get; set; }
}

/// <summary>A sending (and optionally receiving) domain.</summary>
public class Domain
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "domain";

    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;

    /// <summary>
    /// Aggregated verification status: <c>not_started</c>, <c>pending</c>, <c>verified</c>,
    /// <c>partially_verified</c>, <c>partially_failed</c>, <c>failed</c>,
    /// <c>temporary_failure</c>, or <c>revoked</c>.
    /// </summary>
    [JsonPropertyName("status")]
    public string Status { get; set; } = null!;

    [JsonPropertyName("region")]
    public string? Region { get; set; }

    /// <summary>
    /// The DNS zone the registrar manages (e.g. <c>acme.com</c> when Name is
    /// <c>replies.acme.com</c>). Record names should be entered relative to it.
    /// </summary>
    [JsonPropertyName("zone")]
    public string? Zone { get; set; }

    [JsonPropertyName("created_at")]
    public string CreatedAt { get; set; } = null!;

    [JsonPropertyName("records")]
    public List<DomainRecord> Records { get; set; } = new();

    /// <summary>MAIL FROM subdomain (Return-Path); defaults to <c>send</c>.</summary>
    [JsonPropertyName("custom_return_path")]
    public string? CustomReturnPath { get; set; }

    [JsonPropertyName("open_tracking")]
    public bool? OpenTracking { get; set; }

    [JsonPropertyName("click_tracking")]
    public bool? ClickTracking { get; set; }

    /// <summary>Open/click tracking subdomain label (null ⇒ tracking on the shared host).</summary>
    [JsonPropertyName("tracking_subdomain")]
    public string? TrackingSubdomain { get; set; }

    /// <summary>Outbound TLS policy: <c>opportunistic</c> | <c>enforced</c>.</summary>
    [JsonPropertyName("tls")]
    public string? Tls { get; set; }

    [JsonPropertyName("capabilities")]
    public DomainCapabilities? Capabilities { get; set; }

    [JsonPropertyName("tracking_domain")]
    public string? TrackingDomain { get; set; }

    [JsonPropertyName("tracking_verified")]
    public bool? TrackingVerified { get; set; }
}

/// <summary>Payload for creating a domain (POST /domains).</summary>
public class DomainCreateOptions
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;

    [JsonPropertyName("region")]
    public string? Region { get; set; }

    /// <summary>MAIL FROM subdomain (Return-Path); defaults to <c>send</c>.</summary>
    [JsonPropertyName("custom_return_path")]
    public string? CustomReturnPath { get; set; }

    [JsonPropertyName("open_tracking")]
    public bool? OpenTracking { get; set; }

    [JsonPropertyName("click_tracking")]
    public bool? ClickTracking { get; set; }

    /// <summary>Custom tracking host label, e.g. <c>email</c> ⇒ email.&lt;domain&gt;.</summary>
    [JsonPropertyName("tracking_subdomain")]
    public string? TrackingSubdomain { get; set; }

    /// <summary>Outbound TLS policy: <c>opportunistic</c> | <c>enforced</c>.</summary>
    [JsonPropertyName("tls")]
    public string? Tls { get; set; }

    /// <summary>Capabilities to enable, e.g. <c>new() { Receiving = "enabled" }</c>.</summary>
    [JsonPropertyName("capabilities")]
    public DomainCapabilitiesOptions? Capabilities { get; set; }
}

/// <summary>Payload for updating a domain (PATCH /domains/:id).</summary>
public class DomainUpdateOptions
{
    [JsonPropertyName("open_tracking")]
    public bool? OpenTracking { get; set; }

    [JsonPropertyName("click_tracking")]
    public bool? ClickTracking { get; set; }

    [JsonPropertyName("tracking_subdomain")]
    public string? TrackingSubdomain { get; set; }

    /// <summary>Enable/disable the custom open/click tracking host for this domain.</summary>
    [JsonPropertyName("custom_tracking")]
    public bool? CustomTracking { get; set; }

    /// <summary>Outbound TLS policy: <c>opportunistic</c> | <c>enforced</c>.</summary>
    [JsonPropertyName("tls")]
    public string? Tls { get; set; }

    [JsonPropertyName("capabilities")]
    public DomainCapabilitiesOptions? Capabilities { get; set; }
}

/// <summary>The TXT record proving a domain claim.</summary>
public class DomainClaimRecord
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = "TXT";

    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;

    [JsonPropertyName("value")]
    public string? Value { get; set; }

    [JsonPropertyName("ttl")]
    public string? Ttl { get; set; }
}

/// <summary>A claim on a domain already verified by another account.</summary>
public class DomainClaim
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "domain_claim";

    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;

    [JsonPropertyName("domain_id")]
    public string DomainId { get; set; } = null!;

    [JsonPropertyName("region")]
    public string? Region { get; set; }

    [JsonPropertyName("status")]
    public string Status { get; set; } = null!;

    [JsonPropertyName("record")]
    public DomainClaimRecord Record { get; set; } = null!;

    [JsonPropertyName("blocked_reason")]
    public string? BlockedReason { get; set; }

    [JsonPropertyName("failure_reason")]
    public string? FailureReason { get; set; }

    [JsonPropertyName("created_at")]
    public string CreatedAt { get; set; } = null!;

    [JsonPropertyName("expires_at")]
    public string? ExpiresAt { get; set; }
}

/// <summary>Payload for claiming a domain verified elsewhere (POST /domains/claim).</summary>
public class DomainClaimOptions
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;

    [JsonPropertyName("region")]
    public string? Region { get; set; }
}

/// <summary>
/// Result of DNS-provider detection (GET /domains/:id/dns/detect): the provider,
/// its nameservers and the one-click apply methods available.
/// </summary>
public class DnsDetectionResult
{
    [JsonPropertyName("provider")]
    public JsonElement? Provider { get; set; }

    [JsonPropertyName("nameservers")]
    public List<string> Nameservers { get; set; } = new();

    [JsonPropertyName("methods")]
    public List<JsonElement> Methods { get; set; } = new();
}

/// <summary>
/// Namecheap API credentials for one-click DNS apply. Namecheap must have the
/// calling server's IP whitelisted in the account's API settings.
/// </summary>
public class NamecheapDnsCredentials
{
    [JsonPropertyName("apiUser")]
    public string ApiUser { get; set; } = null!;

    [JsonPropertyName("apiKey")]
    public string ApiKey { get; set; } = null!;

    [JsonPropertyName("userName")]
    public string? UserName { get; set; }
}

File: mailblastr-rust/src/services/domains.rs
//! `mailblastr.domains` — verified sending/receiving domains, claims of
//! domains verified elsewhere, and one-click DNS application.

use std::sync::Arc;

use reqwest::Method;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

use crate::client::{page_query, seg, Config};
use crate::types::{ListResponse, ObjectAck, PaginationParams, RemovedResponse, Result};

/// Outbound TLS policy for a domain.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Tls {
    Opportunistic,
    Enforced,
}

/// Whether a domain capability (e.g. receiving) is on.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CapabilityStatus {
    Enabled,
    Disabled,
}

/// One DNS record the domain owner must publish.
#[derive(Debug, Clone, Deserialize)]
pub struct DomainRecord {
    /// `DKIM` | `SPF` | `DMARC` | `Tracking` | `Receiving MX`.
    pub record: String,
    pub name: String,
    #[serde(rename = "type")]
    pub record_type: String,
    pub ttl: String,
    pub status: String,
    pub value: String,
    pub priority: Option<u32>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct DomainCapabilities {
    pub sending: CapabilityStatus,
    pub receiving: CapabilityStatus,
}

/// A sending domain and its DNS records.
#[derive(Debug, Clone, Deserialize)]
pub struct Domain {
    pub object: String,
    pub id: String,
    pub name: String,
    /// Aggregated verification status: `not_started`, `pending`, `verified`,
    /// `partially_verified`, `partially_failed`, `failed`,
    /// `temporary_failure`, or `revoked`.
    pub status: String,
    pub region: String,
    /// The DNS zone the registrar manages (e.g. `acme.com` when `name` is
    /// `replies.acme.com`). Record names should be entered relative to it.
    pub zone: Option<String>,
    pub created_at: String,
    #[serde(default)]
    pub records: Vec<DomainRecord>,
    /// MAIL FROM subdomain (Return-Path); defaults to `send`.
    pub custom_return_path: Option<String>,
    pub open_tracking: Option<bool>,
    pub click_tracking: Option<bool>,
    /// Open/click tracking subdomain label (`None` ⇒ tracking on the shared host).
    pub tracking_subdomain: Option<String>,
    pub tls: Option<Tls>,
    pub capabilities: Option<DomainCapabilities>,
    pub tracking_domain: Option<String>,
    pub tracking_verified: Option<bool>,
}

/// Capabilities to enable on create/update, e.g. receiving.
#[derive(Debug, Clone, Default, Serialize)]
pub struct CapabilitiesInput {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub receiving: Option<CapabilityStatus>,
}

/// Options for `domains.create` (`POST /domains`).
#[derive(Debug, Clone, Default, Serialize)]
pub struct CreateDomainOptions {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub region: Option<String>,
    /// MAIL FROM subdomain (Return-Path); defaults to `send`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_return_path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub open_tracking: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub click_tracking: Option<bool>,
    /// Custom tracking host label, e.g. `email` ⇒ `email.<domain>`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tracking_subdomain: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tls: Option<Tls>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub capabilities: Option<CapabilitiesInput>,
}

impl CreateDomainOptions {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            ..Default::default()
        }
    }

    pub fn with_region(mut self, region: impl Into<String>) -> Self {
        self.region = Some(region.into());
        self
    }

    pub fn with_custom_return_path(mut self, custom_return_path: impl Into<String>) -> Self {
        self.custom_return_path = Some(custom_return_path.into());
        self
    }

    pub fn with_open_tracking(mut self, open_tracking: bool) -> Self {
        self.open_tracking = Some(open_tracking);
        self
    }

    pub fn with_click_tracking(mut self, click_tracking: bool) -> Self {
        self.click_tracking = Some(click_tracking);
        self
    }

    pub fn with_tracking_subdomain(mut self, tracking_subdomain: impl Into<String>) -> Self {
        self.tracking_subdomain = Some(tracking_subdomain.into());
        self
    }

    pub fn with_tls(mut self, tls: Tls) -> Self {
        self.tls = Some(tls);
        self
    }

    /// Enable/disable inbound receiving for this domain.
    pub fn with_receiving(mut self, receiving: CapabilityStatus) -> Self {
        self.capabilities
            .get_or_insert_with(CapabilitiesInput::default)
            .receiving = Some(receiving);
        self
    }
}

/// Options for `domains.update` (`PATCH /domains/:id`).
#[derive(Debug, Clone, Default, Serialize)]
pub struct UpdateDomainOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub open_tracking: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub click_tracking: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tracking_subdomain: Option<String>,
    /// Enable/disable the custom open/click tracking host for this domain.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_tracking: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tls: Option<Tls>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub capabilities: Option<CapabilitiesInput>,
}

impl UpdateDomainOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_open_tracking(mut self, open_tracking: bool) -> Self {
        self.open_tracking = Some(open_tracking);
        self
    }

    pub fn with_click_tracking(mut self, click_tracking: bool) -> Self {
        self.click_tracking = Some(click_tracking);
        self
    }

    pub fn with_tracking_subdomain(mut self, tracking_subdomain: impl Into<String>) -> Self {
        self.tracking_subdomain = Some(tracking_subdomain.into());
        self
    }

    pub fn with_custom_tracking(mut self, custom_tracking: bool) -> Self {
        self.custom_tracking = Some(custom_tracking);
        self
    }

    pub fn with_tls(mut self, tls: Tls) -> Self {
        self.tls = Some(tls);
        self
    }

    pub fn with_receiving(mut self, receiving: CapabilityStatus) -> Self {
        self.capabilities
            .get_or_insert_with(CapabilitiesInput::default)
            .receiving = Some(receiving);
        self
    }
}

/// The TXT record proving ownership during a claim.
#[derive(Debug, Clone, Deserialize)]
pub struct DomainClaimRecord {
    #[serde(rename = "type")]
    pub record_type: String,
    pub name: String,
    pub value: Option<String>,
    pub ttl: String,
}

/// A claim on a domain already verified by another account (takeover path;
/// last verifier wins).
#[derive(Debug, Clone, Deserialize)]
pub struct DomainClaim {
    pub object: String,
    pub id: String,
    pub name: String,
    pub domain_id: String,
    pub region: String,
    pub status: String,
    pub record: DomainClaimRecord,
    pub blocked_reason: Option<String>,
    pub failure_reason: Option<String>,
    pub created_at: String,
    pub expires_at: String,
}

/// Options for `domains.claim` (`POST /domains/claim`).
#[derive(Debug, Clone, Serialize)]
pub struct ClaimDomainOptions {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub region: Option<String>,
}

impl ClaimDomainOptions {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            region: None,
        }
    }

    pub fn with_region(mut self, region: impl Into<String>) -> Self {
        self.region = Some(region.into());
        self
    }
}

/// DNS provider detection result (`GET /domains/:id/dns/detect`).
#[derive(Debug, Clone, Deserialize)]
pub struct DnsDetection {
    pub provider: Value,
    #[serde(default)]
    pub nameservers: Vec<String>,
    /// The one-click apply methods available (Cloudflare token, GoDaddy
    /// key/secret, hosted Domain Connect, panel deep-link).
    #[serde(default)]
    pub methods: Vec<Value>,
}

/// Credentials for the Namecheap one-click DNS apply.
#[derive(Debug, Clone, Serialize)]
pub struct NamecheapDnsOptions {
    #[serde(rename = "apiUser")]
    pub api_user: String,
    #[serde(rename = "apiKey")]
    pub api_key: String,
    #[serde(rename = "userName", skip_serializing_if = "Option::is_none")]
    pub user_name: Option<String>,
}

impl NamecheapDnsOptions {
    pub fn new(api_user: impl Into<String>, api_key: impl Into<String>) -> Self {
        Self {
            api_user: api_user.into(),
            api_key: api_key.into(),
            user_name: None,
        }
    }

    pub fn with_user_name(mut self, user_name: impl Into<String>) -> Self {
        self.user_name = Some(user_name.into());
        self
    }
}

/// `mailblastr.domains`.
#[derive(Clone, Debug)]
pub struct DomainsSvc {
    config: Arc<Config>,
}

impl DomainsSvc {
    pub(crate) fn new(config: Arc<Config>) -> Self {
        Self { config }
    }

    /// Add a domain. `POST /domains`
    pub async fn create(&self, options: CreateDomainOptions) -> Result<Domain> {
        self.config
            .send(self.config.request(Method::POST, "/domains").json(&options))
            .await
    }

    /// Retrieve a domain and its DNS records. `GET /domains/:id`
    pub async fn get(&self, domain_id: &str) -> Result<Domain> {
        let path = format!("/domains/{}", seg(domain_id));
        self.config
            .send(self.config.request(Method::GET, &path))
            .await
    }

    /// List domains. `GET /domains`
    pub async fn list(&self, params: Option<PaginationParams>) -> Result<ListResponse<Domain>> {
        let req = self
            .config
            .request(Method::GET, "/domains")
            .query(&page_query(params.as_ref()));
        self.config.send(req).await
    }

    /// Update tracking/TLS/capabilities. Returns the slim `{ object, id }` ack.
    /// `PATCH /domains/:id`
    pub async fn update(&self, domain_id: &str, options: UpdateDomainOptions) -> Result<ObjectAck> {
        let path = format!("/domains/{}", seg(domain_id));
        self.config
            .send(self.config.request(Method::PATCH, &path).json(&options))
            .await
    }

    /// Trigger DNS verification. Returns the slim `{ object, id }` ack.
    /// `POST /domains/:id/verify`
    pub async fn verify(&self, domain_id: &str) -> Result<ObjectAck> {
        let path = format!("/domains/{}/verify", seg(domain_id));
        self.config
            .send(self.config.request(Method::POST, &path))
            .await
    }

    /// Claim a domain already verified elsewhere. `POST /domains/claim`
    pub async fn claim(&self, options: ClaimDomainOptions) -> Result<DomainClaim> {
        self.config
            .send(
                self.config
                    .request(Method::POST, "/domains/claim")
                    .json(&options),
            )
            .await
    }

    /// Retrieve a domain's claim record. `GET /domains/:id/claim`
    pub async fn get_claim(&self, domain_id: &str) -> Result<DomainClaim> {
        let path = format!("/domains/{}/claim", seg(domain_id));
        self.config
            .send(self.config.request(Method::GET, &path))
            .await
    }

    /// Verify a domain claim. `POST /domains/:id/claim/verify`
    pub async fn verify_claim(&self, domain_id: &str) -> Result<DomainClaim> {
        let path = format!("/domains/{}/claim/verify", seg(domain_id));
        self.config
            .send(self.config.request(Method::POST, &path))
            .await
    }

    /// Detect the domain's DNS provider and available one-click apply methods.
    /// `GET /domains/:id/dns/detect`
    pub async fn detect_dns(&self, domain_id: &str) -> Result<DnsDetection> {
        let path = format!("/domains/{}/dns/detect", seg(domain_id));
        self.config
            .send(self.config.request(Method::GET, &path))
            .await
    }

    /// Apply this domain's DNS records via the Cloudflare API, then
    /// auto-verify. `POST /domains/:id/dns/cloudflare`
    pub async fn apply_cloudflare_dns(&self, domain_id: &str, token: &str) -> Result<Value> {
        let path = format!("/domains/{}/dns/cloudflare", seg(domain_id));
        self.config
            .send(
                self.config
                    .request(Method::POST, &path)
                    .json(&json!({ "token": token })),
            )
            .await
    }

    /// Apply this domain's DNS records via the GoDaddy API, then auto-verify.
    /// `POST /domains/:id/dns/godaddy`
    pub async fn apply_godaddy_dns(
        &self,
        domain_id: &str,
        key: &str,
        secret: &str,
    ) -> Result<Value> {
        let path = format!("/domains/{}/dns/godaddy", seg(domain_id));
        self.config
            .send(
                self.config
                    .request(Method::POST, &path)
                    .json(&json!({ "key": key, "secret": secret })),
            )
            .await
    }

    /// Apply this domain's DNS records via the Namecheap API (existing
    /// records preserved), then auto-verify. Namecheap must have the calling
    /// server's IP whitelisted. `POST /domains/:id/dns/namecheap`
    pub async fn apply_namecheap_dns(
        &self,
        domain_id: &str,
        options: NamecheapDnsOptions,
    ) -> Result<Value> {
        let path = format!("/domains/{}/dns/namecheap", seg(domain_id));
        self.config
            .send(self.config.request(Method::POST, &path).json(&options))
            .await
    }

    /// Delete a domain. `DELETE /domains/:id`
    pub async fn remove(&self, domain_id: &str) -> Result<RemovedResponse> {
        let path = format!("/domains/{}", seg(domain_id));
        self.config
            .send(self.config.request(Method::DELETE, &path))
            .await
    }
}
